src/GACProxyAdmin.sol

```diff
@@ -0,0 +1,113 @@
+// OpenZeppelin Contracts v4.4.1 (proxy/transparent/ProxyAdmin.sol)
+
+pragma solidity ^0.8.0;
+
+import "openzeppelin-contracts/proxy/transparent/TransparentUpgradeableProxy.sol";
+import "./lib/GlobalAccessControlManaged.sol";
+
+/**
+ * @dev This is a version of OZ ProxyAdmin that will use GAC instead of ownable
+ * explanation of why you would want to use this see the documentation for {TransparentUpgradeableProxy}.
+ */
+contract GACProxyAdmin is GlobalAccessControlManaged {
+    bytes32 public constant CONTRACT_GOVERNANCE_ROLE =
+        keccak256("CONTRACT_GOVERNANCE_ROLE");
+
+    /**
+     * @dev Sets the address of the initial implementation, and the deployer account as the owner who can upgrade the
+     * beacon.
+     */
+    function initialize(address _globalAccessControl) public initializer {
+        __GlobalAccessControlManaged_init(_globalAccessControl);
+    }
+
+    /**
+     * @dev Returns the current implementation of `proxy`.
+     *
+     * Requirements:
+     *
+     * - This contract must be the admin of `proxy`.
+     */
+    function getProxyImplementation(TransparentUpgradeableProxy proxy)
+        public
+        view
+        virtual
+        returns (address)
+    {
+        // We need to manually run the static call since the getter cannot be flagged as view
+        // bytes4(keccak256("implementation()")) == 0x5c60da1b
+        (bool success, bytes memory returndata) = address(proxy).staticcall(
+            hex"5c60da1b"
+        );
+        require(success);
+        return abi.decode(returndata, (address));
+    }
+
+    /**
+     * @dev Returns the current admin of `proxy`.
+     *
+     * Requirements:
+     *
+     * - This contract must be the admin of `proxy`.
+     */
+    function getProxyAdmin(TransparentUpgradeableProxy proxy)
+        public
+        view
+        virtual
+        returns (address)
+    {
+        // We need to manually run the static call since the getter cannot be flagged as view
+        // bytes4(keccak256("admin()")) == 0xf851a440
+        (bool success, bytes memory returndata) = address(proxy).staticcall(
+            hex"f851a440"
+        );
+        require(success);
+        return abi.decode(returndata, (address));
+    }
+
+    /**
+     * @dev Changes the admin of `proxy` to `newAdmin`.
+     *
+     * Requirements:
+     *
+     * - This contract must be the current admin of `proxy`.
+     */
+    function changeProxyAdmin(
+        TransparentUpgradeableProxy proxy,
+        address newAdmin
+    ) public virtual onlyRole(CONTRACT_GOVERNANCE_ROLE) {
+        proxy.changeAdmin(newAdmin);
+    }
+
+    /**
+     * @dev Upgrades `proxy` to `implementation`. See {TransparentUpgradeableProxy-upgradeTo}.
+     *
+     * Requirements:
+     *
+     * - This contract must be the admin of `proxy`.
+     */
+    function upgrade(TransparentUpgradeableProxy proxy, address implementation)
+        public
+        virtual
+        onlyRole(CONTRACT_GOVERNANCE_ROLE)
```

Comment:
@dapp-whisperer @Ajand, wouldn't it be a good a idea to have a separate role for this? In case we want to use a Timelock as the only actor allowed to execute upgrades.

Reply:
Same goes for, `upgradeAndCall` and `changeProxyAdmin`

Reply:
@sajanrajdev I think that makes sense